File: .github/workflows/update.py
#!/usr/bin/env python

"""Run with https://docs.astral.sh/uv/guides/scripts:

.. code-block:: shell
    uv run --script .github/workflows/update.py
"""

import json
import subprocess
from pathlib import Path

REPO_DIR = Path(__file__).parent.parent.parent.absolute()
PACKAGE_NAME = "@biomejs/biome"


def main() -> int:
    tags = get_existing_tags()
    versions = get_all_versions()
    missing_tags = get_missing_tags(tags, versions)
    if not missing_tags:
        print("No new versions found")
        return 0
    for tag in sorted(missing_tags):
        print(f"Updating to {tag}")
        update_files(tag)
        stage_commit_and_tag(tag)
    return 0


def get_existing_tags() -> list[str]:
    tags = git("tag", "--list").splitlines()
    return sorted(t for t in tags if t.startswith("v"))


def get_all_versions() -> list[str]:
    all_versions = _get_node_package_versions(PACKAGE_NAME)
    all_versions = [v for v in all_versions if "nightly" not in v]
    return sorted(all_versions)


def get_missing_tags(existing_tags: list[str], all_versions: list[str]) -> set[str]:
    tags = set(existing_tags)
    versions = set(all_versions)
    expected_tags = {f"v{t}" for t in versions}
    return expected_tags - tags


def _get_node_package_versions(package_name: str) -> list[str]:
    cmd = ("npm", "view", package_name, "--json")
    output = json.loads(subprocess.check_output(cmd))
    return output["versions"]


def update_files(tag: str) -> None:
    version = to_version(tag)
    _replace_in_readme(version)
    _replace_in_package_json(version)


def to_version(tag: str) -> str:
    return tag.replace("v", "", 1)


def _replace_in_readme(version: str) -> None:
    readme_file = Path(REPO_DIR, "README.md")
    readme = readme_file.read_text()
    current_version = __get_current_version()
    new_readme = readme.replace(current_version, version)
    readme_file.write_text(new_readme)


def __get_current_version() -> str:
    package = json.loads(Path(REPO_DIR, "package.json").read_text())
    return package["dependencies"][PACKAGE_NAME]


def _replace_in_package_json(version: str) -> None:
    package_json = Path(REPO_DIR, "package.json")
    package = json.loads(package_json.read_text())
    package["dependencies"][PACKAGE_NAME] = version
    package_json.write_text(json.dumps(package, indent=2))


def stage_commit_and_tag(tag: str) -> None:
    git("add", "package.json", "README.md")
    git("commit", "-m", f"MAINT: upgrade to {PACKAGE_NAME} {tag}")
    git("tag", tag)


def git(*cmd: str) -> str:
    output = subprocess.check_output(("git", "-C", REPO_DIR) + cmd)
    return output.decode()


if __name__ == "__main__":
    raise SystemExit(main())

File: .github/workflows/update.js
#!/usr/bin/env node

const util = require("util");
const exec = util.promisify(require("child_process").exec);
const fs = require("fs").promises;
const path = require("path");

const { DEFAULT_BRANCH, GITHUB_ACTOR, GITHUB_ACTOR_ID } = process.env;
const REPO_DIR = path.resolve(__dirname, "..", "..");
const PACKAGE_NAME = "@biomejs/biome";

async function main() {
  const missingVersions = await getMissingVersions();
  if (!missingVersions.length) {
    console.log("No new versions found");
    return;
  }

  await setGitConfig(missingVersions);
  for (const version of missingVersions) {
    console.log(getMessage(version));
    await updateFiles(version);
    await commitTag(version);
  }

  await mergePullRequest(missingVersions);
}

async function getMissingVersions() {
  const allVersions = await getAllVersions();
  const existingVersions = await getExistingVersions();
  const ignoredVersions = await getIgnoredVersions();
  const excludedVersions = new Set([...existingVersions, ...ignoredVersions]);
  return allVersions
    .filter((v) => !excludedVersions.has(v))
    .sort();
}

async function getAllVersions() {
  const allVersions = await getNodePackageVersions(PACKAGE_NAME);
  return allVersions.filter((v) => !v.includes("nightly")).sort();
}

async function getNodePackageVersions(packageName) {
  const { stdout } = await exec(`npm view ${packageName} --json`);
  const output = JSON.parse(stdout);
  return output.versions;
}

async function getIgnoredVersions() {
  const ignoreFile = path.join(REPO_DIR, "ignored-versions.txt");
  const ignores = await fs.readFile(ignoreFile, "utf8");
  return ignores
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line && !line.startsWith("#"));
}

async function getExistingVersions() {
  const existingTags = await git("tag", "--list");
  return existingTags
    .split("\n")
    .map((t) => t.replace("v", "", 1))
    .sort();
}

function getMessage(version) {
  return `build(deps): bump ${PACKAGE_NAME} to ${version}`;
}

async function setGitConfig(versions) {
  const email = `${GITHUB_ACTOR_ID}+${GITHUB_ACTOR}@users.noreply.github.com`;

  await git("config", "user.name", GITHUB_ACTOR);
  await git("config", "user.email", email);
  await git("checkout", DEFAULT_BRANCH);
  await git("checkout", "-b", getBranchName(versions.at(-1)));
}

function getBranchName(version) {
  return `build/bump-${PACKAGE_NAME.replace("/", "-")}-${version}`;
}

async function updateFiles(version) {
  await updateReadme(version);
  await updatePackageJson(version);
  await exec("npm update --package-lock-only");
}

async function updateReadme(version) {
  const readmeFile = path.join(REPO_DIR, "README.md");
  const readme = await fs.readFile(readmeFile, "utf8");
  const newReadme = readme.replace(/rev:\s+\S+/i, `rev: v${version}`);
  await fs.writeFile(readmeFile, newReadme, "utf8");
}

async function updatePackageJson(version) {
  const packageJsonFile = path.join(REPO_DIR, "package.json");
  const packageJson = JSON.parse(await fs.readFile(packageJsonFile, "utf8"));
  packageJson.dependencies[PACKAGE_NAME] = version;
  await fs.writeFile(
    packageJsonFile,
    JSON.stringify(packageJson, null, 2),
    "utf8"
  );
}

async function commitTag(version) {
  const tag = `v${version}`;
  const message = getMessage(version);

  await git("add", "README.md", "package.json", "package-lock.json");
  await git("commit", "--message", `"${message}"`);
  await git("tag", "--annotate", tag, "--message", `"${message}"`);
}

async function mergePullRequest(versions) {
  await git(
    "push",
    "--follow-tags",
    "--set-upstream",
    "origin",
    getBranchName(versions.at(-1))
  );
  await exec(`gh pr create --fill --title "${getMessage(versions.at(-1))}"`);
  await exec("gh pr merge --auto --merge --delete-branch");
}

async function git(...cmd) {
  const { stdout, stderr } = await exec(
    ["git", "-C", REPO_DIR, ...cmd].join(" ")
  );
  if (stderr) console.log(stderr);
  return stdout;
}

main();
